Wrappers/Python/cil/optimisation/operators/WaveletOperator.py

```diff
@@ -0,0 +1,261 @@
+# -*- coding: utf-8 -*-
+#  Copyright 2023 United Kingdom Research and Innovation
+#  Copyright 2023 The University of Manchester
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+# Authors:
+# CIL Developers, listed at: https://github.com/TomographicImaging/CIL/blob/master/NOTICE.txt
+
+import numpy as np
+import pywt  # PyWavelets module
+import warnings
+
+from cil.optimisation.operators import LinearOperator
+from cil.framework import VectorGeometry
+
+
+class WaveletOperator(LinearOperator):
+
+    r'''                  
+        Computes forward or inverse (adjoint) discrete wavelet transform of the input
+
+        Parameters
+        ----------
+        param domain_geometry: cil geometry 
+            Domain geometry for the WaveletOperator
+
+        param range_geometry: cil geometry, optional
```

Comment:
What's the benefit of the user passing a range geometry and it not just being calculated anyway?

Reply:
Minimal.
The only benefit I can see is the user already having a range geometry (for some weird reason) and wanting to use that. This guarantees that the user given geometry is used (or at least tried and then immediately rejected if it's wrong). This also keeps WaveletOperator in line with the other LinearOperators, although I agree this is a very rarely needed functionality.

Reply:
Sounds good - let's leave it as is